### PR TITLE
[ObjC] CF EventEngine client

### PR DESCRIPTION
**Pull request by HannahShiSFB**
@sampajano

### CODE REVIEW

File: src/core/lib/event_engine/cf_engine/cf_engine.cc
Comment (sampajano) [inline src/core/lib/event_engine/cf_engine/cf_engine.cc:113]: there's a null check here, do we need it?
https://github.com/grpc/grpc/blob/1bcaa3e9351a9b00f5ce2a4b68c5e3bf61f6be8d/src/core/lib/event_engine/posix_engine/posix_engine.cc#L301-L312

File: src/core/lib/event_engine/cf_engine/cfstream_endpoint.cc
Comment (drfloob) [inline src/core/lib/event_engine/cf_engine/cfstream_endpoint.cc:250]: I'm a bit lost here, what is actually writing the data? I believe the lockfree event here is triggered when a write finishes, and no write has been initiated yet. Does this work?

File: test/core/event_engine/test_suite/BUILD
Comment (drfloob) [inline test/core/event_engine/test_suite/BUILD:123]: This should be `False`. When `uses_polling = True`, the same test is replicated for every iomgr poller that works on the platform. We don't care about iomgr for these tests.

File: test/core/event_engine/test_suite/cf_event_engine_test.cc
Comment (drfloob) [inline test/core/event_engine/test_suite/cf_event_engine_test.cc:35]: This may mean that the posix engine is used for the `Listener` implementation on whatever platform is running the iOS tests. Is that ok?

Comment (HannahShiSFB) [reply test/core/event_engine/test_suite/cf_event_engine_test.cc:35]: Yes, client test needs a server which we don't have in cf_engine

File: test/core/event_engine/test_suite/BUILD
Comment (HannahShiSFB) [reply test/core/event_engine/test_suite/BUILD:83]: I believe it caused additional tests with polling=none (or something similar) which fails, let me change it back and test it again

File: src/core/lib/event_engine/cf_engine/cfstream_endpoint.cc
Comment (HannahShiSFB) [reply src/core/lib/event_engine/cf_engine/cfstream_endpoint.cc:98]: I can do it here and when CFWriteStreamOpen fails, but not that in open_event_ closure which are executed after this function returns.

File: src/core/lib/event_engine/cf_engine/cf_engine.cc
Comment (HannahShiSFB) [reply src/core/lib/event_engine/cf_engine/cf_engine.cc:113]: I can add the check but it won't help anything as
1. unlike that in PosixEnginePollerManager, here executor_ is created in engine constructor so will not be null 
2. if it were null then closures cannot run so it will stuck which is not better than crash where we have stack trace
3. can we instead somehow annotate it not null or assert here?

File: src/core/lib/event_engine/cf_engine/cf_engine.cc
Comment (HannahShiSFB) [reply src/core/lib/event_engine/cf_engine/cf_engine.cc:83]: I don't think there is an ios specific one, I can use the same event engine category for logging

File: src/core/lib/event_engine/cf_engine/cf_engine.cc
Comment (drfloob) [reply src/core/lib/event_engine/cf_engine/cf_engine.cc:98]: I see, thanks for explaining. It should be valid to have an EventEngine instance that does not support connection cancellation, I imagine there may be cases where things like gRPC shutdown takes longer. We can try it without that support.

File: src/core/lib/event_engine/cf_engine/cfstream_endpoint.cc
Comment (HannahShiSFB) [reply src/core/lib/event_engine/cf_engine/cfstream_endpoint.cc:116]: Mmm, it sounds that I'll have to do some tracking or reference counting after all, even if we don't support cancel connection, to prevent from this being deleted because of connection error while the timeout timer is being fired.
I'll need sometime on this.

File: src/core/lib/event_engine/cf_engine/cftype_unique_ref.h
Comment (sampajano) [inline src/core/lib/event_engine/cf_engine/cftype_unique_ref.h:26]: @drfloob Hey AJ I wonder there are existing utilities like this in the grpc codebase?

This class seems quite generic to me so i'm hoping we could reuse if there exists one. Thanks!

Comment (drfloob) [reply src/core/lib/event_engine/cf_engine/cftype_unique_ref.h:26]: It looks like this class provides some sugar specifically for `CFTypeRef` objects (implicit conversions and output-parameter specializations). It looks to me like a `std::unique_ptr` could do this job, albeit without the sugar on top. I'm mostly ambivalent about it, but we don't have anything `CFTypeRef`-specific in the core codebase.

Comment (ctiller) [reply src/core/lib/event_engine/cf_engine/cftype_unique_ref.h:26]: This looks like `template <typename T> using CFTypeUniqueRef = std::unique_ptr<T, CFRelease>;`

Comment (HannahShiSFB) [reply src/core/lib/event_engine/cf_engine/cftype_unique_ref.h:26]: I tries with something like
```
template <typename T>
using CFTypeUniqueRef =
    std::unique_ptr<std::remove_cvref_t<decltype(*std::declval<T>())>,
                    decltype(& CFRelease)>;
```
Which is very difficult to use because:
1. Unlike typical unique_ptr use case, `T` (CFStringRef, CFErrorRef, etc.) is the pointer type thus the complex alias
2. Deleter (&CFRelease) needs to be passed everywhere creating a pointer
2. I don't do dereference (operator* and operator->) but only convert back and forth when using with CF APIs.
3. I need `operator&` for using as output parameter

Comment (HannahShiSFB) [reply src/core/lib/event_engine/cf_engine/cftype_unique_ref.h:26]: Alternative to this, I can also try warp a unique pointer for auto release then add implicit conversions and `operator&` if you prefer. It'll be something like
```
template <typename T>
class CFTypeUniqueRef {
public:
  // conversions, operator&
private:
  std::unique_ptr<std::remove_cvref_t<decltype(*std::declval<T>())>,
                    decltype(& CFRelease)> ref_;
};
```

**Comment by HannahShiSFB** [comment]
@drfloob Hi AJ, I think I have addressed all the comments, could you take a look and review it again? Thank you

File: src/core/lib/event_engine/cf_engine/cf_engine.cc
Comment (drfloob) [reply src/core/lib/event_engine/cf_engine/cf_engine.cc:159]: That's a fine justification. Thanks.

**Comment by drfloob** [comment]
@HannahShiSFB it looks like the `EventEngineClientTest.MultipleIPv6ConnectionsToOneOracleListenerTest` test crashes on Mac with a SIGBUS error.  See the `Basic Tests C/C++ MacOS ` check. If it reproduces locally, the command will be something like `bazel test --config=dbg --test_env=GRPC_EXPERIMENTS=event_engine_client --test_env=GRPC_VERBOSITY=debug --test_filter='*MultipleIPv6ConnectionsToOneOracleListenerTest*' //test/core/event_engine/test_suite:cf_event_engine_test`.

File: src/core/lib/gprpp/time.h
Comment (drfloob) [inline src/core/lib/gprpp/time.h:41]: I suggest reverting this change, explicit atomic operations are preferred in this codebase, and we use `atomic::exchange` a fair amount. This change removes an explicit call, and adds an implicit load and store.

File: test/core/event_engine/test_suite/BUILD
Comment (drfloob) [reply test/core/event_engine/test_suite/BUILD:85]: Did you re-test this?

File: src/core/BUILD
Comment (drfloob) [inline src/core/BUILD:2069]: Is this necessary? `port_platform.h` should set this automatically for the right platforms. And further, we would have to be careful to ensure that no general target accidentally depends on `cf_event_engine` (even if it's unused), because the defines here will propagate to every rule that depends on it ... on every platform. https://bazel.build/reference/be/c-cpp#cc_binary.defines

Comment (HannahShiSFB) [reply src/core/BUILD:2069]: That's what I wondered. I initially tried to set it to the test target only but grpc_cc_test doesn't have it.
Still trying other configurations for running tests.
Besides, it appears some tests don't pass with GRPC_CFSTREAM with iomgr, still looking

**Comment by HannahShiSFB** [comment]
> the EventEngineClientTest.MultipleIPv6ConnectionsToOneOracleListenerTest test crashes on Mac with a SIGBUS error

Passing locally, investigating

File: src/core/lib/gprpp/time.h
Comment (drfloob) [reply src/core/lib/gprpp/time.h:41]: It looks like https://github.com/grpc/grpc/pull/32204 fixes the bug as well, and only uses one `exchange`. That will likely land today. Thanks for catching the bug!

File: test/core/event_engine/test_suite/BUILD
Comment (HannahShiSFB) [inline test/core/event_engine/test_suite/BUILD:83]: This fails with `GRPC_POLL_STRATEGY=none`

https://source.cloud.google.com/results/invocations/4cead0a6-9484-4fd7-80ec-2fac56cf8746/targets/github%2Fgrpc%2Frun_tests%2Fcpp_macos_dbg_native/tests

CC: @drfloob

**Comment by HannahShiSFB** [comment]
@drfloob I noticed that https://github.com/grpc/grpc/pull/32253 completely disabled experimental on mobile platforms. How should I enable cf event engine in our ios tests in "src/objective-c/tests/"?

PS, I cannot enable it in mac tests as many of the tests are using socket files instead of ip:port, which is not supported by cfstream.

**Comment by drfloob** [comment]
> How should I enable cf event engine in our ios tests in "src/objective-c/tests/"?

Pinging @ctiller for considerations into a partial rollback of https://github.com/grpc/grpc/pull/32253. The PR has no description, I presume this solved some actual bloat problems.

File: test/core/event_engine/test_suite/BUILD
Comment (drfloob) [reply test/core/event_engine/test_suite/BUILD:83]: I see, thanks.

**Comment by HannahShiSFB** [comment]
Android CI failure is not related.
All review comments addressed.
Will enable cf event engine in iOS tests in a separate PR once experimental features can be enabled.

**Comment by drfloob** [comment]
@HannahShiSFB after chatting with @ctiller, we'll need another solution for enabling experiments on iOS - the way it's set up today resulted in 10s of MB of binary bloat on iOS, which caused some problems internally.

We can set it up differently for iOS, though, and instead of a runtime flag, add the ability to build gRPC with `bazel test --config=event_engine_client [targets]`. We would need to set up a separate CI job for it though, since it would build gRPC from scratch twice, and we'd risk making presubmit CI suite take much longer otherwise. @sampajano do you have experience setting up new pull request CI jobs?

File: src/core/lib/event_engine/cf_engine/cfstream_endpoint.cc
Comment (sampajano) [reply src/core/lib/event_engine/cf_engine/cfstream_endpoint.cc:189]: Borrowing this thread: I wonder if `kCFStreamEventEndEncountered` should also trigger `write_event_` (which triggers a write), or `kCFStreamEventCanAcceptBytes` alone is sufficient?

File: src/core/lib/event_engine/cf_engine/cfstream_endpoint.cc
Comment (sampajano) [inline src/core/lib/event_engine/cf_engine/cfstream_endpoint.cc:324]: Am I reading it correctly, that this is recursively scheduling a write, if it cannot finish in one `CFWriteStreamWrite` call?

I wonder if this is susceptible to a race condition if `CFStreamEndpoint::Write` is called again and now there are 2 scheduled writes?

File: src/core/lib/event_engine/cf_engine/cftype_unique_ref.h
Comment (sampajano) [inline src/core/lib/event_engine/cf_engine/cftype_unique_ref.h:65]: I asked ChatGPT(3.5) if there's any bug in this class (i have to say i love i can just do that 😂) and it says if you add a `if (cf_type_ref_ != other)` check it could prevent a potential issue if both are the same.. i don't know if it truly matters but FYI.. 😄

**Review by sampajano — COMMENTED** [review]
@HannahShiSFB Hi Hannah, thanks SO MUCH for working on this change!! Looks fantastic!! Just a bunch of relatively minor questions and comments inline! Thanks again for working on this!!! ❤️

**Comment by sampajano** [comment]
> @HannahShiSFB after chatting with @ctiller, we'll need another solution for enabling experiments on iOS - the way it's set up today resulted in 10s of MB of binary bloat on iOS, which caused some problems internally.
> 
> We can set it up differently for iOS, though, and instead of a runtime flag, add the ability to build gRPC with `bazel test --config=event_engine_client [targets]`. We would need to set up a separate CI job for it though, since it would build gRPC from scratch twice, and we'd risk making presubmit CI suite take much longer otherwise. @sampajano do you have experience setting up new pull request CI jobs?

@drfloob Sorry i forgot to reply here. As we discussed offline earlier, i do not have previous experiences with new CI jobs for grpc, but would be happy to work with you on figuring that out! Will reach out once this PR is merged. Thanks!

File: src/core/lib/event_engine/cf_engine/cf_engine.cc
Comment (HannahShiSFB) [reply src/core/lib/event_engine/cf_engine/cf_engine.cc:101]: Not really as deadline_timer is a plain struct, copy and move are the same. So I guess it's more like a documentation of moving the ownership into the lambda then.

File: src/core/lib/event_engine/cf_engine/cf_engine.cc
Comment (HannahShiSFB) [reply src/core/lib/event_engine/cf_engine/cf_engine.cc:114]: I don't know what "valid but unspecified" state mean, but in any case we should not (and we didn't) access the moved-from object.

File: src/core/lib/event_engine/cf_engine/cf_engine.cc
Comment (HannahShiSFB) [reply src/core/lib/event_engine/cf_engine/cf_engine.cc:148]: This will trigger the shutdown of open_event_ in CFStreamEndpoint, then the on_connect callback, it'll be deleted there.

File: src/core/lib/event_engine/cf_engine/cfstream_endpoint.cc
Comment (HannahShiSFB) [reply src/core/lib/event_engine/cf_engine/cfstream_endpoint.cc:36]: Don't know, this is from the old cfstream implementation

File: src/core/lib/event_engine/cf_engine/cfstream_endpoint.cc
Comment (HannahShiSFB) [reply src/core/lib/event_engine/cf_engine/cfstream_endpoint.cc:185]: In my test it's always read becomes ready first then write is ready. For any reason if read is not ready yet, it's unlikely to be an issue as client will send request first.
Old cfstream_handle have it in both read and write which means which ever happens first triggers the event which seems worked fine as well.

open_event is connected to the on_connect callback, it'll either be set ready (connected) or shutdown with a status (connection failed or canceled or timed out)

File: src/core/lib/event_engine/cf_engine/cfstream_endpoint.cc
Comment (HannahShiSFB) [reply src/core/lib/event_engine/cf_engine/cfstream_endpoint.cc:324]: "There can be at most one outstanding write per Endpoint at any given time", this is how c-core uses endpoint.

File: src/core/lib/event_engine/cf_engine/cftype_unique_ref.h
Comment (HannahShiSFB) [reply src/core/lib/event_engine/cf_engine/cftype_unique_ref.h:65]: It's actually right, to prevent self-reset

File: src/core/lib/event_engine/cf_engine/cfstream_endpoint.cc
Comment (HannahShiSFB) [reply src/core/lib/event_engine/cf_engine/cfstream_endpoint.cc:189]: I think kCFStreamEventCanAcceptBytes alone should be sufficient, the old cfstream does this way, borrowed from there

File: src/core/lib/iomgr/endpoint_cfstream.cc
Comment (sampajano) [reply src/core/lib/iomgr/endpoint_cfstream.cc:243]: ah thanks for explaining!

File: src/core/lib/event_engine/cf_engine/cf_engine.cc
Comment (sampajano) [reply src/core/lib/event_engine/cf_engine/cf_engine.cc:101]: i see! i'm ok with this if this is consistent with the rest of the grpc code base (if it is). :)

File: src/core/lib/event_engine/cf_engine/cf_engine.cc
Comment (sampajano) [reply src/core/lib/event_engine/cf_engine/cf_engine.cc:114]: aha ok! i guess this is fine.. (i've seen this used in other classes too so i assume this is how it's typically used.. :))

File: src/core/lib/event_engine/cf_engine/cfstream_endpoint.cc
Comment (sampajano) [reply src/core/lib/event_engine/cf_engine/cfstream_endpoint.cc:36]: ok sg thanks :)

File: src/core/lib/event_engine/cf_engine/cf_engine.cc
Comment (sampajano) [reply src/core/lib/event_engine/cf_engine/cf_engine.cc:148]: aha ok! just want to double check. thanks for explaining!

File: src/core/lib/event_engine/cf_engine/cfstream_endpoint.cc
Comment (sampajano) [reply src/core/lib/event_engine/cf_engine/cfstream_endpoint.cc:324]: aha ok got it! Good to note! Thanks for clarifying!

File: src/core/lib/event_engine/cf_engine/cfstream_endpoint.cc
Comment (sampajano) [reply src/core/lib/event_engine/cf_engine/cfstream_endpoint.cc:185]: aha ok! thanks for explaining!

Could you maybe put a little bit comment somewhere to make this assumption more clear? Thanks!

**Comment by HannahShiSFB** [comment]
All comments addressed except the [grpc_event_engine_endpoint_data_trace](https://github.com/grpc/grpc/pull/32077#discussion_r1148519440) one.

File: src/core/lib/event_engine/cf_engine/cfstream_endpoint.cc
Comment (drfloob) [reply src/core/lib/event_engine/cf_engine/cfstream_endpoint.cc:275]: No, it's not necessary, probably not useful on iOS.

**Comment by drfloob** [comment]
> As we discussed offline, i do not have previous experiences with new CI jobs for grpc, but would be happy to work with you on figuring that out! Will reach out once this PR is merged. Thanks!

@sampajano I put together a CI test that should work for you, https://github.com/grpc/grpc/pull/32782. And please refer to our offline chat for the other piece required to land it. I'll be afk until next week, hopefully this is sufficient to get you started with EventEngine CI tests on iOS.

**Comment by sampajano** [comment]
@drfloob Thanks SOO much for helping putting together a change here!! I have asked @HannahShiSFB to help testing them. 😃 

@HannahShiSFB Maybe you could provide an update once you have a good idea on whether the tests are working, and if not, why. Thanks! 😃

**Comment by sampajano** [comment]
> All comments addressed except the [grpc_event_engine_endpoint_data_trace](https://github.com/grpc/grpc/pull/32077#discussion_r1148519440) one.

@HannahShiSFB Thanks Hannah!!

So, after you've done fixing it, could you go over the comments again and ensure all open comments are either:

1) Addressed AND resolved
2) Addressed AND responded, and waiting for confirmation from reviewer?

I've still seeing some comments like [this one](https://github.com/grpc/grpc/pull/32077/files#r1152609051) and [this one](https://github.com/grpc/grpc/pull/32077/files#r1152171360) which are unclear whether they're addressed or not.

Thanks!!

**Comment by HannahShiSFB** [comment]
Enabled cf engine in #32791 in ios and test passes.
Tests were flaky because c-core may delete an endpoint at any time, fixed by [use ref counting for CFStreamEndpointImpl](https://github.com/grpc/grpc/pull/32077/commits/ca4115094d1d5595b00e7baa2525ae18dc077901)
Stress run InteropTestsLocalCleartext 500 times and passing.

**Comment by HannahShiSFB** [comment]
> I've still seeing some comments like [this one](https://github.com/grpc/grpc/pull/32077/files#r1152609051) and [this one](https://github.com/grpc/grpc/pull/32077/files#r1152171360) which are unclear whether they're addressed or not.

All comments addressed, added comment for first one, no change to the second (follow the old implementation)

**Comment by sampajano** [comment]
> All comments addressed ...
> 
> Failed CI "Distribution Tests Python Windows" is not related

Woohoo! Thanks so much for all the hard work on pulling this through, Hannah!! 😃

@drfloob Hey AJ do you want to take a last pass? If not, this can be merged? Thanks!

**Review by drfloob — APPROVED** [review]
Looks good. Thanks for the hard work, Hannah!

**Comment by sampajano** [comment]
> Looks good. Thanks for the hard work, Hannah!

Thanks for the hard work indeed, Hannah!! Please feel free to merge the PR whenever you feel ready to! Thanks!! 🎉 🥳 😃

---

@drfloob Also thanks so much for being super helpful during the review process too, AJ! 😃

**Pull request merged**
